zarr/storage.py

```diff
@@ -1877,6 +1878,197 @@ def __delitem__(self, key):
             self._invalidate_value(key)
 
 
+class SQLiteStore(MutableMapping):
+    """Storage class using SQLite.
+
+    Parameters
+    ----------
+    path : string
+        Location of database file.
+    **kwargs
+        Keyword arguments passed through to the `sqlite3.connect` function.
+
+    Examples
+    --------
+    Store a single array::
+
+        >>> import zarr
+        >>> store = zarr.SQLiteStore('data/array.sqldb')
+        >>> z = zarr.zeros((10, 10), chunks=(5, 5), store=store, overwrite=True)
+        >>> z[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+
+    Store a group::
+
+        >>> store = zarr.SQLiteStore('data/group.sqldb')
+        >>> root = zarr.group(store=store, overwrite=True)
+        >>> foo = root.create_group('foo')
+        >>> bar = foo.zeros('bar', shape=(10, 10), chunks=(5, 5))
+        >>> bar[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+    """
+
+    def __init__(self, path, **kwargs):
+        import sqlite3
+
+        kwargs.setdefault('timeout', 5.0)
+        kwargs.setdefault('cached_statements', 100)
+
+        # normalize path
+        if path != ':memory:':
+            path = os.path.abspath(path)
+
+        # store properties
+        self.path = path
+        self.kwargs = kwargs
+
+        # open database
+        self.db = sqlite3.connect(
+            self.path,
+            detect_types=0,
+            isolation_level=None,
```

Comment:
Is there a reason to force `isolation_level=None`? We might want to let the user determine the isolation level so they can override the default and manage transactions themselves if they want to.

Reply:
This ensures the default SQLite behavior. To clarify, I mean the C library's default behavior and not Python's. Namely that every transaction is [automatically committed]( https://www.sqlite.org/lockingv3.html#transaction_control ) once completed. IMHO this seems desirable from a multiprocessing standpoint, but I could be missing things.

Reply:
Automatically committing will mean that there is a commit after every key/value put, and that seems like a sensible default. But I can imagine a situation where a user is writing a whole array, or some large region of an array, or maybe even multiple arrays, and they want that entire task to either succeed or fail, and so want to have some control over transactions. I don't have a concrete use case for this myself, but it would seem better at this stage to allow this flexibility, i.e., set `isolation_level=None` as default but allow the user to override via kwargs.

Reply:
Sure that makes sense. This includes a custom implementation of the `MutableMapping.update` method, which adds multiple key-value pairs in a single commit. Have had some success with this strategy for different use cases outside of Zarr. So tried adding it for storing data here. We might considering implementing `update` for other database style stores here and making more use of these `update` methods of stores where we want this behavior.

Reply:
Cool. This actually touches on something I've been thinking about, having a standard method on store classes for batch operations, and making use of it in zarr. This would open up the possibility for stores to do a number of interesting things they can't do currently. Using `MutableMapping.update()` is a nice route.

Reply:
Glad you like it.

We've been using a `MutableMapping` subclass to check-in intermediate Dask arrays as Zarr Arrays on disk. So `update` has been pretty handy when storing multiple Dask arrays that share intermediates; thus, benefiting from being computed together.

Would be great to hear about other use cases you have in mind for storing multiple key-value pairs.

Reply:
Cool. I should make a separate issue to discuss possibilities, thought I had done a while back but can't find it now. Basically when modifying a region of an array that spans more than one chunk, zarr could pass this through to the storage layer as a single batch operation, rather than as separate operations for each key/value pair. The store could then do various things. E.g., if it was a cloud store, and the underlying cloud service offered a batch API, then the store could use that to improve performance. Cloud stores could also do other things like combine multiple chunks into a single storage object, which again could reduce the number of API requests being made. Other types of stores might implement some transactional behaviour as you've done here, so that there is some protection against an array modification involving multiple chunks being left half-done. Lots of possibilities.

Reply:
Sounds very interesting. Looking forward to discussing in the issue :)